cypress-io/cypress
# chore: record event gql mutation

- Closes https://github.com/cypress-io/cypress-internal/issues/632

### Additional details


| NOTE | 
| -- |
| This PR is dependent on the Cloud PR: https://github.com/cypress-io/cypress-services/pull/6165  |

Adding additional way to capture event data in Cypress Cloud using a GQL mutation rather than separate endpoints.  This allows for recording different events with different payloads without continuing to expand the fields on the existing event types.

This PR introduces one new event called LocalTestCounts.  This passes 4 count values for specs and tests within a project as well as breaking out separate counts for files that match the example e2e paths.

NOTE: The count for tests is a rudimentary algorithm that is just looking for a regex in the files.  It is not meant to be totally accurate for when tests are created through functions, but rather give an estimate of the amount of effort put into writing tests.

#### How event is triggered

The `EventCollectorSource` watches two events that are triggered by other parts of the code (`authChange` and `configChange`).  I added two methods to subscribe (`subscribeToRawEvent`) and unsubscribe(`unsubscribeToRawEvent`) to those events directly in the `DataEmitterActions` without having to deal with the `AsyncIterator` that the GQL subscriptions need to operate.  

### Steps to test


To monitor the outgoing events, you can turn on DEBUG statements for:
* `cypress:data-context:sources:EventCollectorSource`
* `cypress:data-context:actions:EventCollectorActions`

```
DEBUG=cypress:data-context:sources:EventCollectorSource,cypress:data-context:actions:EventCollectorActions yarn dev
```

Preconditions that must be met for LocalTestCounts event to be sent:
* must be logged in
* must have loaded a project config
* must not have sent an event for this project within the last 30 days
 * this being tracked through a new saved setting in project state called `lastTestCountsEvent`.  This is saved as a timestamp for when the last event was sent

Scenarios to test:
* Project does not have `lastTestCountsEvent` in the `state.json` file for the project or the saved timestamp is older than 30 days
  * Local project has no spec files
    * Should see counts of all zeros
  * Local project is not connected to a cloud project
    * Should NOT see a `projectSlug` value in the event
  * Project has specs matching the path and name for the example scaffolded e2e specs 
    * Can use the `cypress-example-kitchensink` project as a test since that is where all the examples come from or scaffold a new project
      * Expect for the `example` counts for specs and tests to match the `total` counts
* Project has a  `lastTestCountsEvent` timestamp saved in the `state.json` file that is less than 30 days old
  * Should see a debug message from `cypress:data-context:sources:EventCollectorSource` that says "will not send" and shows that "hasBeenSentLast30Days" is `false`

Example steps of when to see the events be triggered:
* In global mode
  * after selecting a project and then a testing type, this will trigger the `configChange` event
* in regular project mode
  * after opening the App and selecting a testing type, this will trigger the `configChange` event
* in either mode
  * log out of the cloud
  * open a project and choose a testing type
  * log in, this will trigger the 'authChange' event   



### How has the user experience changed?


N/A

### PR Tasks
